File: TypeScript/206. Reverse Linked List.ts
/*
Given the head of a singly linked list, reverse the list, and return the reversed list.

206. Reverse Linked List
*/
class ListNode {
    val: number;
    next: ListNode | null;

    constructor(val?: number, next?: ListNode | null) {
        this.val = val ?? 0;
        this.next = next ?? null;
    }
}

// Function to build a tree from an array representation
function buildList(arr: (number | null)[]): ListNode | null {
    const n = arr.length;
    if (n === 0 || arr[0] === null) { return null; }

    const root = new ListNode(arr[0]);
    const queue: ListNode[] = [root];
    let i = 1;

    while (i < n) {
        const node = queue.shift() as ListNode;

        if (i < n && arr[i] !== null) {
            node.next = new ListNode(arr[i]!);
            queue.push(node.next);
        }
        i++;
    }
    return root;
}
function listToArray(head) {
    let arr = [];
    while (head) {
        arr.push(head.val);
        head = head.next;
    }
    return arr;
}

function reverseList(head: ListNode | null): ListNode | null {
    if (head === null) { return head; }
    let prev: ListNode | null = null;
    let current: ListNode | null = head;
    while (current !== null) {
        const next: ListNode | null = current.next;
        current.next = prev;
        prev = current
        current = next;
    }
    return prev;    
};

//const head = buildList([1,2,3,4,5]); // Expect [5,4,3,2,1]
//const head = buildList([1,2]); // Expect [2,1]
const head = buildList([]); // Expect []
console.log(listToArray(reverseList(head))); // Expect true

File: Swift/206. Reverse Linked List.swift
/*
Given the head of a singly linked list, reverse the list, and return the reversed list.
    
206. Reverse Linked List    
*/
public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
}

public func buildList(_ arr: [Int]) -> ListNode? {
    if arr.isEmpty { return nil }
    let n = arr.count
    let root = ListNode(arr[0])
    var current = root
    
    for i in 1..<n {
        current.next = ListNode(arr[i])
        current = current.next!
    }
    return root    
}

public func listToArray(_ head: ListNode?) -> [Int] {
    var arr: [Int] = []
    var current = head
    
    while let node = current {
        arr.append(node.val)
        current = node.next
    }
    return arr
}

class Solution {
    func reverseList(_ head: ListNode?) -> ListNode? {
        var prev: ListNode? = nil
        var current = head
        
        while let node = current {
            let next = node.next
            node.next = prev
            prev = node
            current = next
        }
        
        return prev
    }
}

let solution = Solution()
if let linkedList = buildList([1, 2, 3, 4, 5]) {
    let reversedList = solution.reverseList(linkedList)
    print(listToArray(reversedList)) // Expect: [5, 4, 3, 2, 1]
} 
else {
    print("Empty list")
}

if let linkedList = buildList([1, 2]) {
    let reversedList = solution.reverseList(linkedList)
    print(listToArray(reversedList)) // Expect: [2, 1]
} 
else {
    print("Empty list")
}

if let linkedList = buildList([]) {
    let reversedList = solution.reverseList(linkedList)
    print(listToArray(reversedList)) // This won’t execute as the list is empty
} 
else {
    print("Empty list") // Expect: "Empty list"
}
